Add --max-warnings cli flag (fixes #2939)

alexilyaev opened the pull request:
Resolves: https://github.com/stylelint/stylelint/issues/2939

A CLI flag useful in CI to exit with non zero code on warnings count greater than a specific number.  
Usually `defaultSeverity: 'warning'` is set and `--max-warnings 0` in CLI.  
That way devs don't see errors in development, but CI fails build if there are any warnings.

<img width="673" alt="screen shot 2017-10-05 at 24 45 50" src="https://user-images.githubusercontent.com/480469/31203028-dcefea52-a96d-11e7-920c-b7bc359ed3ad.png">

alexilyaev commented:
@jeddy3 Can you guide me how to write a test for it?  
I worked with Jest in the past, just need a direction.

alexilyaev commented:
Running `npm run lint` gives me this:
```
...

/Users/alex/www/Node/OSS/stylelint/system-tests/systemTestUtils.js
   1:13  error  Extra semicolon         semi
   3:28  error  Extra semicolon         semi
   4:29  error  Extra semicolon         semi
   5:40  error  Extra semicolon         semi
   8:52  error  Extra semicolon         semi
  12:50  error  Extra semicolon         semi
  16:22  error  Extra semicolon         semi
  17:51  error  Extra semicolon         semi
  23:59  error  Extra semicolon         semi
  28:54  error  Extra semicolon         semi
  30:25  error  Extra semicolon         semi
  31:5   error  Extra semicolon         semi
  35:26  error  Extra semicolon         semi
  43:14  error  Missing trailing comma  comma-dangle
  44:2   error  Extra semicolon         semi

✖ 27123 problems (27123 errors, 0 warnings)
```

I installed deps with Yarn, if it makes a difference.

alexilyaev commented:
@jeddy3 Thanks.

Btw, do I need to do anything in `standalone.js`?  
What is it used for?

jeddy3 commented:
> Btw, do I need to do anything in standalone.js? What is it used for?

It's the main guts of the [Node API](https://github.com/stylelint/stylelint/blob/master/docs/user-guide/node-api.md#options) and where the [returned object](https://github.com/stylelint/stylelint/blob/master/docs/user-guide/node-api.md#the-returned-promise) is [constructed](https://github.com/stylelint/stylelint/blob/master/lib/standalone.js#L233-L250). A `maxWarningsExceeded` property could be added to this returned object, which `cli.js` can use.

alexilyaev commented:
@jeddy3 So, I've moved the logic to `standalone.js`, adding the `maxWarningsExceeded` property, updated the docs and added a test file.

Although I'm not sure what would be the use case for using it through the Node API.

This PR is ready for a review.

jeddy3 commented on lib/cli.js:
```diff
@@ -473,11 +496,22 @@ Promise.resolve()
+      const foundWarnings = linted.maxWarningsExceeded.foundWarnings;
+
+      process.stdout.write(
+        `Stylelint found too many warnings:${EOL}` +
```
Can we tweak the message to make it consistent [with how other errors (e.g. "Invalid Option" ones)](https://cloud.githubusercontent.com/assets/808227/14970067/063e5802-10be-11e6-8774-6557d80c443b.png) are formatted, please? e.g. :

```js
  process.stdout.write(
        chalk.red(`Max warnings exceeded: `) +
          `${foundWarnings} found. ` +
          chalk.dim(`${maxWarnings} allowed${EOL}${EOL}`)
      );
```

alexilyaev replied:
@jeddy3 Looks good, updated.

jeddy3 commented:
@CAYdenberg If you have a moment, would you be able to advise of those flow warnings?

CAYdenberg commented:
Done. @alexilyaev hope you don't mind me appending to your PR. Thanks for your contribution.

jeddy3 commented:
@CAYdenberg Thanks!

alexilyaev commented:
@CAYdenberg Not at all, thanks.

So, merging any time soon?

jeddy3 commented:
@stylelint/core Can anyone spare a moment to cast a 2nd pair of eyes over this?

jeddy3 commented:
Change log:

> -   Added: `--max-warnings` CLI flag ([#2942](https://github.com/stylelint/stylelint/pull/2942)).
